# Analyzing Sales Funnel for a Student Hostel Accomodation Chain 
### Summary Table For The Funnel

*______________________________________________________________________________________________________________________________*

### Introduction :-
  
  This notebook focuses on the exploration of sales funnel analysis. The initial step in the analysis involves describing the datasets and finding relationships between them. By identifying common features, we merge the datasets to obtain a comprehensive dataset that encompasses all the stages of the sales funnel, from pre-verification to onboarding, for each city. This combined dataset provides valuable insights into the transfer of leads between different stages. To download a summary table of the cumulative dataset, simply execute all the lines of code.

##### Pre-requisites 
Loading all the essential packages, collections, and libraries to proceed with the analysis.

In [1]:
import pandas as pd
import numpy as np

### Data Collection :-
In this analysis, we will utilize the dataset provided by Strent AH Pvt Limited (SAPL).

In [7]:
status = pd.read_csv(r"\Users\HP\Downloads\status_table.csv", encoding= 'unicode_escape')
cities = pd.read_csv(r"\Users\HP\Downloads\city_table.csv", encoding= 'unicode_escape')
All_leads = pd.read_csv(r"\Users\HP\Downloads\lead_visit_table.csv", encoding= 'unicode_escape')
onboarding_leads = pd.read_csv(r"\Users\HP\Downloads\hostel_table.csv", encoding= 'unicode_escape')

### Exploratory Data Analysis :-
In this phase, we will perform exploratory data analysis to gain a better understanding of the datasets. We will start by displaying the dataframes or the first few rows of the status, cities, visited_leads, and onboarding_leads datasets to examine their contents. Additionally, we will explore the unique values present in each column of the dataframes to gather insights into all possible values. Furthermore, we will identify common columns between the datasets and merge them when necessary using these common columns.

In [3]:
status

substatus_id             substatus_name        status
0              1                     signup  pre verified
1              2             mobile_updated  pre verified
2              3              email_updated  pre verified
3              4           id_proof_updated  pre verified
4              5  hostel_preference_updated  pre verified
5              6        Police_verification  pre verified
6              7             email_verified       Verifed
7              8                otp_verifed       Verifed
8              9          id_proof_approved       Verifed
9             10              Verified_lead       Verifed
10            11               call_pending     connected
11            12                  connected     connected
12            13            Visit_Scheduled       Visited
13            14            Visit_cancelled       Visited
14            15                 Visit_done       Visited
15            16           Visit_reschduled       Visited
16            17               Visit_missed       Visited
17            18        Aggrement initiated     Agreement
18            19           Neotiation_stage     Agreement
19            20           Aggrement shared     Agreement
20            21           Aggrement_Signed     Agreement
21            22     Agreement edit request     Agreement
22            23            Payment Recvied     Agreement
23            24       onboarding initiated     onboarded
24            25  hostel_preference_alloted     onboarded
25            26            student_shifted     onboarded

The status dataframe assigns different substatus_ids corresponding to different stages.

In [4]:
cities

city_id        city
0       11       Delhi
1       12       Noida
2       13   Bangalore
3       14    Gurugram
4       15      Mumbai
5       16        Pune
6       17   Hyderabad
7       18   Ahmedabad
8       19  Chandigarh
9       20     Kolkata

The cities dataframe assigns a unique city_id to each city to ensure their distinct identification.

In [8]:
All_leads.head()

id        created_at  lead_id  sustatus_id  hostel_id  cityid
0   1  19-03-2023 02:50     1001            1        NaN      14
1   2  19-03-2023 13:13     1001            2        NaN      14
2   3  20-03-2023 23:07     1001            3        NaN      14
3   4  21-03-2023 04:20     1001            4        NaN      14
4   5  21-03-2023 22:31     1001            5     3116.0      14

In [6]:
onboarding_leads.head()

id  hostel_id  city_id
0   1       3101     11.0
1   2       3102     11.0
2   3       3103     11.0
3   4       3104     11.0
4   5       3105     11.0

The All_leads dataframe provides a description of leads that were generated or converted at each stage across different cities and dates. Meanwhile, the onboarding_leads dataframe presents a list of leads who have successfully completed the agreement process and are either about to be onboarded or have already been onboarded to their respective hostels.

Upon examining the dataframes, we observe that some rows in the All_leads dataframe do not have hostel_ids assigned to them. This indicates that these leads have not yet reached or are far from the agreement stage. The allocation of a hostel_id is done for leads who have completed the agreement stage.

#### Merging Dataframes

In this step, we will perform the merging of dataframes. By combining the relevant datasets using common columns, we can create a unified dataset that consolidates the information from different stages and cities. This merged dataset will enable us to gain comprehensive insights and conduct further analysis on the sales funnel process for SAPL.

It appears that the substatus_id in the All_leads dataframe corresponds to the substatus_id in the status dataframe. Since a direct comparison of individual columns between different dataframes is not feasible, we will merge them using an outer join. This merging approach ensures that all rows are combined, and any uncommon values are replaced with NaN values, allowing for a comprehensive analysis of the data.

In [11]:
status_of_all_leads = pd.merge(status, All_leads, left_on='substatus_id', right_on='sustatus_id', how='outer')

In [14]:
status_of_all_leads.head()

substatus_id substatus_name        status   id        created_at  lead_id  \
0             1         signup  pre verified    1  19-03-2023 02:50     1001   
1             1         signup  pre verified   27  13-03-2023 00:50     1002   
2             1         signup  pre verified   53  03-03-2023 08:47     1003   
3             1         signup  pre verified   79  11-03-2023 06:54     1004   
4             1         signup  pre verified  105  10-03-2023 11:40     1005   

   sustatus_id  hostel_id  cityid  
0            1        NaN      14  
1            1        NaN      14  
2            1        NaN      11  
3            1        NaN      14  
4            1        NaN      17

In [13]:
status_of_all_leads.shape

(8779, 9)

After merging the status and All_leads dataframes into a new dataframe called status_of_all_leads, we can observe the first 5 rows of status_of_all_leads seems like the sustatus_id is the same as the substatus_id, we cannot confirm this assumption since the displayed subset represents only a small portion of the entire dataframe, which contains 8779 rows. To confirm the consistency of the sustatus_id and substatus_id throughout the merged dataframe :-

In [15]:
#Compare the two columns whether merging was correct or not
are_equal = status_of_all_leads['substatus_id'] == status_of_all_leads['sustatus_id']

# Check if all values are True
all_equal = are_equal.all()

# Print the result
print(all_equal)

True


This confirmation indicates that the values of substatus_id in each row are indeed equal to the corresponding values of sustatus_id. We can now delete the sustatus_id column, as it is essentially a duplicate of the substatus_id column.

In [16]:
status_of_all_leads.drop('sustatus_id', axis=1, inplace=True)

Upon further examination of all the dataframes, it has come to our attention that the All_leads dataframe does not provide information about the cities to which the leads belong. In order to obtain the corresponding city names, we will merge the merged dataframe (status_of_all_leads) with the cities dataframe. This merge will be performed by matching the cityid of the status_of_all_leads dataframe with the city_id of the cities dataframe. The result of this merge will be a new dataframe called status_of_all_leads_with_cities, which will provide us with the necessary information about the cities associated with the leads.

In [22]:
status_of_all_leads_with_cities = pd.merge(status_of_all_leads, cities, left_on='cityid', right_on='city_id', how='outer')

In [23]:
status_of_all_leads_with_cities.head()

substatus_id substatus_name        status   id        created_at  lead_id  \
0             1         signup  pre verified    1  19-03-2023 02:50     1001   
1             1         signup  pre verified   27  13-03-2023 00:50     1002   
2             1         signup  pre verified   79  11-03-2023 06:54     1004   
3             1         signup  pre verified  235  08-03-2023 22:18     1010   
4             1         signup  pre verified  261  01-03-2023 03:55     1011   

   hostel_id  cityid  city_id      city  
0        NaN      14     14.0  Gurugram  
1        NaN      14     14.0  Gurugram  
2        NaN      14     14.0  Gurugram  
3        NaN      14     14.0  Gurugram  
4        NaN      14     14.0  Gurugram

We observe that the cityid is identical to the city_id. However, to fully confirm this,

In [19]:
#Compare the two columns whether merging was correct or not
are_equal = status_of_all_leads_with_cities['cityid'] == status_of_all_leads_with_cities['city_id']

# Check if all values are True
all_equal = are_equal.all()

# Print the result
print(all_equal)

False


The cityid and city_id are not the same, which contradicts the merge operation. To investigate the reason behind this inconsistency, we will display the unique values of cityid and city_id to identify the variations between them.

In [20]:
status_of_all_leads_with_cities['cityid'].unique()

array([14, 11, 17, 16, 15,  0, 13, 20, 19, 18, 12], dtype=int64)

In [21]:
status_of_all_leads_with_cities['city_id'].unique()

array([14., 11., 17., 16., 15., nan, 13., 20., 19., 18., 12.])

The unique values are also same for cityid and city_id. To know why they are not same we will create a new dataframe which is subset of status_of_all_leads_with_cities as rows_with_mismatched_cities which contains rows whose cityid is not same as city_id

In [24]:
rows_with_mismatched_cities = status_of_all_leads_with_cities[status_of_all_leads_with_cities['cityid'] != status_of_all_leads_with_cities['city_id']]
rows_with_mismatched_cities.head()

substatus_id substatus_name        status    id        created_at  \
5552             1         signup  pre verified   209  17-03-2023 21:37   
5553             1         signup  pre verified  1067  06-03-2023 14:34   
5554             1         signup  pre verified  1093  30-03-2023 03:29   
5555             1         signup  pre verified  1119  10-03-2023 21:07   
5556             1         signup  pre verified  1769  21-03-2023 00:24   

      lead_id  hostel_id  cityid  city_id city  
5552     1009        NaN       0      NaN  NaN  
5553     1042        NaN       0      NaN  NaN  
5554     1043        NaN       0      NaN  NaN  
5555     1044        NaN       0      NaN  NaN  
5556     1069        NaN       0      NaN  NaN

Now, let's display the unique elements of cityid and city_id separately to gain a better understanding of the factors that differentiate these columns.

In [25]:
rows_with_mismatched_cities['cityid'].unique()

array([0], dtype=int64)

In [26]:
rows_with_mismatched_cities['city_id'].unique()

array([nan])

The difference between the two columns is that empty values in cityid are represented as 0, while empty values in city_id remain empty. To address this inconsistency, we will fill the empty values of status_of_all_leads_with_cities column city_id with 0.

In [27]:
status_of_all_leads_with_cities['city_id'] = status_of_all_leads_with_cities['city_id'].fillna(0)

Now, we will compare cityid and city_id once again to verify if the solution has resolved the discrepancy.

In [28]:
#Compare the two columns whether merging was correct or not
are_equal = status_of_all_leads_with_cities['cityid'] == status_of_all_leads_with_cities['city_id']

# Check if all values are True
all_equal = are_equal.all()

# Print the result
print(all_equal)

True


Finally, our cityid values are now matching with city_id. Therefore, we can remove the cityid column to eliminate duplicates, reduce storage space, and avoid any potential confusion.

In [29]:
status_of_all_leads_with_cities.drop('cityid', axis=1, inplace=True)

In [30]:
status_of_all_leads_with_cities.head()

substatus_id substatus_name        status   id        created_at  lead_id  \
0             1         signup  pre verified    1  19-03-2023 02:50     1001   
1             1         signup  pre verified   27  13-03-2023 00:50     1002   
2             1         signup  pre verified   79  11-03-2023 06:54     1004   
3             1         signup  pre verified  235  08-03-2023 22:18     1010   
4             1         signup  pre verified  261  01-03-2023 03:55     1011   

   hostel_id  city_id      city  
0        NaN     14.0  Gurugram  
1        NaN     14.0  Gurugram  
2        NaN     14.0  Gurugram  
3        NaN     14.0  Gurugram  
4        NaN     14.0  Gurugram

This appears to be satisfactory. Let's now have a closer look at the 'city_id' and 'city' columns and identify the unique combinations of values.

In [33]:
unique_values = status_of_all_leads_with_cities[['city_id', 'city']].drop_duplicates()
unique_values

city_id        city
0        14.0    Gurugram
1673     11.0       Delhi
3494     17.0   Hyderabad
4082     16.0        Pune
4834     15.0      Mumbai
5552      0.0         NaN
6440     13.0   Bangalore
7635     20.0     Kolkata
8030     19.0  Chandigarh
8144     18.0   Ahmedabad
8380     12.0       Noida

All the values in the city_id column appear to be fine, except for the cases where cityid = 0 and city is NULL. These instances are not ideal, especially considering that the main objective of this notebook is to summarize the distribution of leads for each city.To eliminate the entries with unknown cityid and city, we will remove those records from the dataset.

In [34]:
status_of_all_leads_with_cities = status_of_all_leads_with_cities[status_of_all_leads_with_cities['city_id'] != 0]

In [36]:
unique_values = status_of_all_leads_with_cities[['city_id', 'city']].drop_duplicates()
unique_values

city_id        city
0        14.0    Gurugram
1673     11.0       Delhi
3494     17.0   Hyderabad
4082     16.0        Pune
4834     15.0      Mumbai
6440     13.0   Bangalore
7635     20.0     Kolkata
8030     19.0  Chandigarh
8144     18.0   Ahmedabad
8380     12.0       Noida

We have successfully removed the empty values, effectively cleaning our data. Now, we can proceed with constructing our Summary table for the funnel.

### Getting Started with Summary Table :-

 To construct the Summary table for the funnel, we need to organize the columns in chronological order based on the substatus_id, representing the different stages. Each stage should be uniquely identified by its corresponding stage name. We will start by assigning the column name 'pre-verified' to represent substatus_id 1 to 6. This will be followed by the 'Verified' column, representing substatus_id 7 to 10, and so on until the 'onboarding' stages, which correspond to substatus_id 24 to 26. These column names will be stored in an array called 'column_names', which will later become the columns of the Summary table.

Next, we will collect all the city values and store them in an array called 'city_values', which will become the rows of the Summary table.

For the values in the Summary table, we need to individually gather the status names and store them in an array called 'status_values'. These status names will be used to fill the maximum possible count of each status corresponding to each city in the Summary table.

Finally, we will fill in the columns of stage numbers by counting the occurrences of each substatus_id for each city.

In [37]:
column_names = ['pre verified'] + [f'[Stage {i}]' for i in range(1, 7)] + ['Verifed'] + [f'[Stage {i}]' for i in range(7, 11)] + ['connected'] + [f'[Stage {i}]' for i in range(11, 13)] + ['Visited'] + [f'[Stage {i}]' for i in range(13, 18)] + ['Agreement'] + [f'[Stage {i}]' for i in range(18, 24)] + ['onboarded'] + [f'[Stage {i}]' for i in range(24, 27)]

# Unique city values from df['city']
city_values = status_of_all_leads_with_cities['city'].unique()
status_values = status_of_all_leads_with_cities['status'].unique()
substatus_id_values = status_of_all_leads_with_cities['substatus_id'].unique()

Now, we will create the Summary table by using the arrays 'column_names' and 'city_values' as the column headers and row headers, respectively.

In [38]:
updated_status_of_all_leads_with_cities = pd.DataFrame(columns=column_names, index=city_values)
column = updated_status_of_all_leads_with_cities.columns

To verify if the headers of the summary table are assigned correctly, you can perform the following check.

In [42]:
print(updated_status_of_all_leads_with_cities.columns)

Index(['pre verified', '[Stage 1]', '[Stage 2]', '[Stage 3]', '[Stage 4]',
       '[Stage 5]', '[Stage 6]', 'Verifed', '[Stage 7]', '[Stage 8]',
       '[Stage 9]', '[Stage 10]', 'connected', '[Stage 11]', '[Stage 12]',
       'Visited', '[Stage 13]', '[Stage 14]', '[Stage 15]', '[Stage 16]',
       '[Stage 17]', 'Agreement', '[Stage 18]', '[Stage 19]', '[Stage 20]',
       '[Stage 21]', '[Stage 22]', '[Stage 23]', 'onboarded', '[Stage 24]',
       '[Stage 25]', '[Stage 26]'],
      dtype='object')


In [41]:
print(updated_status_of_all_leads_with_cities)

           pre verified [Stage 1] [Stage 2] [Stage 3] [Stage 4] [Stage 5]  \
Gurugram            NaN       NaN       NaN       NaN       NaN       NaN   
Delhi               NaN       NaN       NaN       NaN       NaN       NaN   
Hyderabad           NaN       NaN       NaN       NaN       NaN       NaN   
Pune                NaN       NaN       NaN       NaN       NaN       NaN   
Mumbai              NaN       NaN       NaN       NaN       NaN       NaN   
Bangalore           NaN       NaN       NaN       NaN       NaN       NaN   
Kolkata             NaN       NaN       NaN       NaN       NaN       NaN   
Chandigarh          NaN       NaN       NaN       NaN       NaN       NaN   
Ahmedabad           NaN       NaN       NaN       NaN       NaN       NaN   
Noida               NaN       NaN       NaN       NaN       NaN       NaN   

           [Stage 6] Verifed [Stage 7] [Stage 8]  ... [Stage 18] [Stage 19]  \
Gurugram         NaN     NaN       NaN       NaN  ...        NaN        N

To populate the values in the Summary table, we will follow these steps:-

 1.We will match the 'city' names from the dataframe 'status_of_all_leads_with_cities' with the row headers of the Summary table.

 2.For the two different categories of columns, namely 'status' and 'substatus_id', from the dataframe 'status_of_all_leads_with_cities', we will use an if condition. If the 'status' matches with a column header in the Summary table, we will count all the records in the dataframe 'status_of_all_leads_with_cities' where the 'city' corresponds to the row header and the 'status' matches the column header.

 3.In case the if condition fails, we will count the occurrences where the 'city' matches the row header and the column header has the format '[Stage <number>]', where the number should match the substatus_id.

By following these steps, we will populate the appropriate counts in the values of the Summary table.

In [43]:
for city in city_values:
    for col in column:
        if col in status_values:
            for status in status_values:
                count = status_of_all_leads_with_cities[(status_of_all_leads_with_cities['city'] == city) & (status_of_all_leads_with_cities['status'] == status)]['lead_id'].nunique()
                updated_status_of_all_leads_with_cities.at[city, status] = count
           
        else:
            for substatus_id in substatus_id_values:
                count_1 = status_of_all_leads_with_cities[(status_of_all_leads_with_cities['city'] == city) & (status_of_all_leads_with_cities['substatus_id'] == substatus_id)].shape[0]
                updated_status_of_all_leads_with_cities.at[city, f'[Stage {substatus_id}]'] = count_1

Now, let's display the table with its updated values.

In [44]:
updated_status_of_all_leads_with_cities

pre verified [Stage 1] [Stage 2] [Stage 3] [Stage 4] [Stage 5]  \
Gurugram            120       120       118       115       113       110   
Delhi               125       125       123       120       115       109   
Hyderabad            38        38        37        35        33        33   
Pune                 60        60        60        59        54        51   
Mumbai               50        50        49        47        46        45   
Bangalore            81        81        80        77        74        70   
Kolkata              27        27        26        26        26        25   
Chandigarh            9         9         9         9         9         8   
Ahmedabad            20        20        20        18        17        17   
Noida                29        29        28        28        27        27   

           [Stage 6] Verifed [Stage 7] [Stage 8]  ... [Stage 18] [Stage 19]  \
Gurugram         105     101       101        93  ...         43         39   
Delhi            105     100       100        96  ...         50         45   
Hyderabad         32      32        32        31  ...         18         16   
Pune              49      48        48        43  ...         15         10   
Mumbai            41      39        39        38  ...         20         19   
Bangalore         67      63        63        62  ...         34         31   
Kolkata           24      24        24        22  ...         10         10   
Chandigarh         8       7         7         6  ...          1          1   
Ahmedabad         17      15        15        14  ...          3          2   
Noida             25      23        23        22  ...         11          8   

           [Stage 20] [Stage 21] [Stage 22] [Stage 23] onboarded [Stage 24]  \
Gurugram           31         26         22         17        14         14   
Delhi              44         39         34         25        19         19   
Hyderabad          12         11         10          9         9          9   
Pune                9          7          6          5         5          5   
Mumbai             16         13         12         10         9          9   
Bangalore          29         25         24         16        15         15   
Kolkata             7          6          5          4         4          4   
Chandigarh          1          1          1          0         0          0   
Ahmedabad           1          1          1          1         1          1   
Noida               6          5          4          2         1          1   

           [Stage 25] [Stage 26]  
Gurugram            9          5  
Delhi              15         10  
Hyderabad           6          2  
Pune                3          1  
Mumbai              6          4  
Bangalore           8          2  
Kolkata             3          3  
Chandigarh          0          0  
Ahmedabad           1          1  
Noida               1          0  

[10 rows x 32 columns]

We can observe that as the stage number increases, the values in the table descend. This progressive decrease in values from the pre-verified stage to the verified stage and ultimately to the onboarding stage accurately depicts the trend of the sales funnel. Therefore, our approach to frame the summary table is appropriate and effectively represents the sales funnel process.

For the final output, we should set the header values to be displayed as true.

In [46]:
# Add 'City/Funnel Stages' column
updated_status_of_all_leads_with_cities.insert(0, 'City/Funnel Stages', city_values)

# Add row values
updated_status_of_all_leads_with_cities.loc[:, 'City/Funnel Stages'] = city_values

# Set 'City/Funnel Stages' as index
updated_status_of_all_leads_with_cities.set_index('City/Funnel Stages', inplace=True)

Here is the final overview of our Summary table.

In [47]:
updated_status_of_all_leads_with_cities

pre verified [Stage 1] [Stage 2] [Stage 3] [Stage 4]  \
City/Funnel Stages                                                        
Gurugram                    120       120       118       115       113   
Delhi                       125       125       123       120       115   
Hyderabad                    38        38        37        35        33   
Pune                         60        60        60        59        54   
Mumbai                       50        50        49        47        46   
Bangalore                    81        81        80        77        74   
Kolkata                      27        27        26        26        26   
Chandigarh                    9         9         9         9         9   
Ahmedabad                    20        20        20        18        17   
Noida                        29        29        28        28        27   

                   [Stage 5] [Stage 6] Verifed [Stage 7] [Stage 8]  ...  \
City/Funnel Stages                                                  ...   
Gurugram                 110       105     101       101        93  ...   
Delhi                    109       105     100       100        96  ...   
Hyderabad                 33        32      32        32        31  ...   
Pune                      51        49      48        48        43  ...   
Mumbai                    45        41      39        39        38  ...   
Bangalore                 70        67      63        63        62  ...   
Kolkata                   25        24      24        24        22  ...   
Chandigarh                 8         8       7         7         6  ...   
Ahmedabad                 17        17      15        15        14  ...   
Noida                     27        25      23        23        22  ...   

                   [Stage 18] [Stage 19] [Stage 20] [Stage 21] [Stage 22]  \
City/Funnel Stages                                                          
Gurugram                   43         39         31         26         22   
Delhi                      50         45         44         39         34   
Hyderabad                  18         16         12         11         10   
Pune                       15         10          9          7          6   
Mumbai                     20         19         16         13         12   
Bangalore                  34         31         29         25         24   
Kolkata                    10         10          7          6          5   
Chandigarh                  1          1          1          1          1   
Ahmedabad                   3          2          1          1          1   
Noida                      11          8          6          5          4   

                   [Stage 23] onboarded [Stage 24] [Stage 25] [Stage 26]  
City/Funnel Stages                                                        
Gurugram                   17        14         14          9          5  
Delhi                      25        19         19         15         10  
Hyderabad                   9         9          9          6          2  
Pune                        5         5          5          3          1  
Mumbai                     10         9          9          6          4  
Bangalore                  16        15         15          8          2  
Kolkata                     4         4          4          3          3  
Chandigarh                  0         0          0          0          0  
Ahmedabad                   1         1          1          1          1  
Noida                       2         1          1          1          0  

[10 rows x 32 columns]

We have successfully created the Summary table for the funnel. To utilize this table, you can download the Excel file associated with it below

In [49]:
updated_status_of_all_leads_with_cities.to_excel(r"\Users\HP\Downloads\D2_Output_of_Funnel_Chart_Report.xlsx", index=True)